CONTRIBUTING.md

```diff
@@ -0,0 +1,149 @@
+
+You should see some output like:
+```terminal
+make test
+running 3 tests
+test instance::tests::test_maybe_open_stdio ... ok
+test instance::wasitest::test_delete_after_create ... ok
+test instance::wasitest::test_wasi ... ok
+```
+
+### End to End tests
+
+The e2e test run on [k3s](https://k3s.io/) and [kind](https://kind.sigs.k8s.io/).  A test image is built using [oci-tar-builder](./crates/oci-tar-builder/) and is loaded onto the clusters.  This test image is not pushed to an external registry so be sure to use the Makefile targets to build the image and load it on the cluster.
+
+The deployment file in [test/k8s/Dockerfile](./test/k8s/Dockerfile) is run and verified that it deploys and runs successfully.  To execute the e2e tests in either k8s or k3s:
+
+```
+make test/k8s-<runtime>
+make test/k3s-<runtime>
+```
+
+OCI Wasm image requires containerd 1.7.7+ and can tested with:
+
+```
+make test/k8s-oci-<runtime>
+```
+
+### Building the test image
+
+This builds a [wasm application](crates/wasi-demo-app/) and packages it in an OCI format:
+
+```
+make test-image
+```
+
+## Code style
+
+We use `Cargo fmt` and `clippy`  for most linting rules.  Use the `check` command in the makefile to verify your code matches the expected style.
+
+```
+make check
+```
+
+You can auto-fix most styles using 
+
+```
+make fix
+```
+
+## Updating protobuf files
+
+Ensure protoc and dev tools have been installed.  Run `make build` or to just generate the protos:
+
+```
+cargo build -p containerd-shim-wasm --no-default-features --features generate_bindings
+```
+
+## Adding new features
+
+Most features will likely have most of the code in the `containerd-shim-wasm` project and a few runtime specific addtions to each runtime shim.  The general expectation is that the feature should be added to all runtimes. We will evaluate on a case by case basis exceptions, where runtimes may not be able to support a given feature or requires changes that make it hard to review.  In those cases we it may make sense to implement in follow up PR's for other runtimes.
+
+A tip for developing a new feature is to implement it and test it with one runtime you are familiar with then add it to all the runtimes.  This makes it easier to test and iterate before making changes across all the runtimes.
+
+## Adding new shims
+
+We welcome new shims, though you can also host them in your own repositories as well and use the `coantainerd-shim-wasm` crate.  We recognize that the project is moving fast to having them in this repository can reduce the need for changes as well for discoverability.  
+
+Please open an issue before submitting a PR for discussion to make sure it is a good fit.  As a general rule, we want shims to be adopting WASI spec. Since we are not experts in every runtime we also need a commitment from the runtime owners to contribute to maintenance of the shim.
+
+## Removing Shims
+
+If a shim in not maintained for 6 months it will be subject to removal.  Before removal:
```

Comment:
Is ambiguous to me if this is saying the "shim" itself is not maintained, or the "runtime" underneath the shim is not maintainer for 6 months.

Reply:
I am not sure how to define this.  Open to wording and we can discuss more in the meeting next week.

I don't think it is just the runtime its self (but that is a good thing to call out here).  But also want to avoid a situation if there is a bug in the runtime that surfaces in our test suite or need changes to runtime to support a feature but no one is around to help resolve quickly even if the runtime is actively maintained.

Reply:
I've taken another shot at this